src/main/java/it/gov/pagopa/payhub/activities/service/IuvService.java

```diff
@@ -0,0 +1,109 @@
+package it.gov.pagopa.payhub.activities.service;
+
+import it.gov.pagopa.payhub.activities.dao.IuvSequenceNumberDao;
+import it.gov.pagopa.payhub.activities.dto.OrganizationDTO;
+import it.gov.pagopa.payhub.activities.exception.ValueNotValidException;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+/**
+ * Service class exposing methods related to IUV handling.
+ */
+@Lazy
+@Service
+@Slf4j
+public class IuvService {
+
+  private static final String PROGRESSIVO_PADDING_ZEROES_SMALL = "%011d";
+  public static final String AUX_DIGIT = "3";
+
+  private final String informationSystemId;
+
+  private final IuvSequenceNumberDao progressiviVersamentoDao;
+
+  public IuvService(@Value("${iuv.informationSystemId:00}") String informationSystemId,
+                    IuvSequenceNumberDao iuvSequenceNumberDao) {
+    this.informationSystemId = informationSystemId;
+    this.progressiviVersamentoDao = iuvSequenceNumberDao;
+  }
+
+
+  /**
+   * Generate a valid and unique IUV given the organization entity.
+   * @param org the organization for which to generate the IUV
+   * @return the generated IUV
+   */
+  public String generateIuv(OrganizationDTO org){
+    StringBuilder iuvBuilder = new StringBuilder();
+    iuvBuilder.append(org.getApplicationCode());
+    iuvBuilder.append(informationSystemId);
+    long paymentIndex = progressiviVersamentoDao.getNextIuvSequenceNumber(org.getIpaCode());
+    if(paymentIndex<1){
+      log.error("invalid payment index returned for org[{}/{}]: {}", org.getIpaCode(), org.getOrgFiscalCode(), paymentIndex);
+      throw new ValueNotValidException("invalid payment index");
+    }
+    iuvBuilder.append(String.format(PROGRESSIVO_PADDING_ZEROES_SMALL, paymentIndex));
```

Comment:
let's move paymentIndex (also padded) building logic on a separate method, in order to let more readable the main method

Reply:
done